File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        def countLessEqual(matrix, mid, smaller, larger):
            count, n = 0, len(matrix)
            row, col = n - 1, 0

            while row >= 0 and col < n:
                if matrix[row][col] > mid:
                    # if the number is greater than mid, keep the track of
                    # the smallest number greater than mid
                    larger = min(larger, matrix[row][col])
                    row -= 1
                else:
                    smaller = max(smaller, matrix[row][col])
                    col += 1
                    count += row + 1
                    
            return count, smaller, larger
            
        n = len(matrix)
        start, end = matrix[0][0], matrix[n - 1][n - 1]
        while start < end:
            mid = (end - start) / 2 + start
            smaller, larger = matrix[0][0], matrix[n - 1][n - 1]
            
            count, smaller, larger = countLessEqual(matrix, mid, smaller, larger)
            
            if count == k: return smaller
            if count < k: 
                start = larger
            else:
                end = smaller
        return start

File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.cpp
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        int n = matrix.size();
        int start = matrix[0][0], end = matrix[n - 1][n - 1];
        int mid, smaller, larger, count;
        
        while (start < end) {
            mid = start + (end - start) / 2;
            smaller = matrix[0][0];
            larger = matrix[n - 1][n - 1];
            
            count = countLessEqual(matrix, mid, smaller, larger);
            
            if (k == count) return smaller;
            // search for higher part
            if (k > count) start = larger;
            // search for lower part
            else end = smaller;
        }
        return start;
    }
private:
    int countLessEqual(vector<vector<int>>& matrix, int& mid, int& smaller, int& larger) {
        // only count the size of the left part
        int count = 0, n = matrix.size();
        int row = n - 1, col = 0;
        
        while (row >= 0 && col < n) {
            // right half of the array
            if (matrix[row][col] > mid) {
                larger = min(larger, matrix[row--][col]);
            }
            else {
                smaller = max(smaller, matrix[row][col++]);
                count += row + 1;
            }
        }
        return count;
    } 
};